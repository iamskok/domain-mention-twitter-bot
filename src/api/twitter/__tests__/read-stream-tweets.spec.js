import EventEmitter from 'events';
import readStreamTweets from '../read-stream-tweets';
import twitUserAuth from '../../../services/twit-user-auth';
import tweetsMock from '../../../mocks/tweets';
import getPostTitle from '../../../utils/get-post-title';

const mockedEmitter = new EventEmitter();
const firestoreMock = {
  data: {},
  async setTweet(postTitle, tweet) {
    this.data[postTitle] = tweet;
  },
  async getTweet(postTitle) {
    return Promise.resolve(this.data[postTitle]);
  },
};

const expectedSnapshot = (tweetMock, postTitle, done) => {
  setTimeout(() => {
    mockedEmitter.emit('tweet', tweetMock);

    setTimeout(async () => {
      expect(firestoreMock.getTweet(postTitle)).resolves.toMatchSnapshot();
      mockedEmitter.removeAllListeners('tweet');
      done();
    }, 1000);
  }, 100);
};

describe('readStreamTweets', () => {
  it('returns `javascript` post title tweet with URL entities, which is not a quote or retweet', (done) => {
    jest.spyOn(twitUserAuth, 'stream')
      .mockImplementation(() => mockedEmitter);

    const url = 'https://example.com/blog/javascript';
    const postTitle = getPostTitle(url);

    readStreamTweets(postTitle, firestoreMock.setTweet.bind(firestoreMock));

    expectedSnapshot(tweetsMock[8], postTitle, done);
  });

  it('returns `react` post title tweet with URL entities, which is not a quote or retweet', (done) => {
    jest.spyOn(twitUserAuth, 'stream')
      .mockImplementation(() => mockedEmitter);

    const url = 'https://example.com/blog/react';
    const postTitle = getPostTitle(url);

    readStreamTweets(postTitle, firestoreMock.setTweet.bind(firestoreMock));

    expectedSnapshot(tweetsMock[9], postTitle, done);
  });

  it('returns undefined for quoted tweet', (done) => {
    jest.spyOn(twitUserAuth, 'stream')
      .mockImplementation(() => mockedEmitter);

    const url = 'https://example.com/blog/go';
    const postTitle = getPostTitle(url);

    readStreamTweets(postTitle, firestoreMock.setTweet.bind(firestoreMock));

    expectedSnapshot(tweetsMock[10], postTitle, done);
  });

  it('returns undefined for retweeted tweet', (done) => {
    jest.spyOn(twitUserAuth, 'stream')
      .mockImplementation(() => mockedEmitter);

    const url = 'https://example.com/blog/typescript';
    const postTitle = getPostTitle(url);

    readStreamTweets(postTitle, firestoreMock.setTweet.bind(firestoreMock));

    expectedSnapshot(tweetsMock[11], postTitle, done);
  });

  it('returns undefined for tweet without URL entities', (done) => {
    jest.spyOn(twitUserAuth, 'stream')
      .mockImplementation(() => mockedEmitter);

    readStreamTweets(undefined, firestoreMock.setTweet.bind(firestoreMock));

    expectedSnapshot(tweetsMock[12], undefined, done);
  });
});
