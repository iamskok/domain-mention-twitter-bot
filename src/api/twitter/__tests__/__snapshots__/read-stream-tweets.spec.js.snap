// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readStreamTweets returns \`javascript\` post title tweet with URL entities, which is not a quote or retweet 1`] = `
Object {
  "created_at": "Wed Oct 10 05:40:24 +0000 2020",
  "entities": Object {
    "urls": Array [
      Object {
        "expanded_url": "https://example.com/blog/javascript",
      },
    ],
  },
  "id_str": "88888888888888888888",
  "in_reply_to_status_id_str": null,
  "quoted_status_id_str": null,
  "quotes": null,
  "replies": null,
  "text": "1st tweet with URL - http://t.co/12345",
  "user": Object {
    "name": undefined,
    "profile_image_url_https": undefined,
    "screen_name": "Mark",
  },
}
`;

exports[`readStreamTweets returns \`react\` post title tweet with URL entities, which is not a quote or retweet 1`] = `
Object {
  "created_at": "Wed Oct 10 06:40:24 +0000 2020",
  "entities": Object {
    "urls": Array [
      Object {
        "expanded_url": "https://example.com/blog/react",
      },
    ],
  },
  "id_str": "030303030303030303",
  "in_reply_to_status_id_str": null,
  "quoted_status_id_str": null,
  "quotes": null,
  "replies": null,
  "text": "2nd tweet with URL - http://t.co/54321",
  "user": Object {
    "name": undefined,
    "profile_image_url_https": undefined,
    "screen_name": "Joe",
  },
}
`;

exports[`readStreamTweets returns undefined for quoted tweet 1`] = `undefined`;

exports[`readStreamTweets returns undefined for retweeted tweet 1`] = `undefined`;

exports[`readStreamTweets returns undefined for tweet without URL entities 1`] = `undefined`;
